### DISCUSSION TITLE
Back noise reduction for OCR

### DISCUSSION THREAD

**Discussion by jmspaggi** [ImageMagick/ImageMagick#7209] [Help]
### ImageMagick version

6.9.11-60

### Operating system, version and so on

Debian 12.5

### Description

Hi,

I have the following picture that I'm trying to "clean" to run Tesseract on it. 

![image](https://github.com/ImageMagick/ImageMagick/assets/3017827/c5c8987c-4c15-41f5-b9ec-61a71fc2b8a3)

I have looked at threads here and there and so far this is what I have done:
convert output1.jpg -colorspace Gray output2.jpg
convert output2.jpg -channel RGB -negate output3.jpg
convert output3.jpg -white-threshold 50%% -transparent white output4.jpg
convert output4.jpg -negate -lat 20x20+10% -negate -despeckle output5.jpg
tesseract output5.jpg stdout

I know I can chain them all on a single command but this helped me to see what each task was doing.
It outputs this:
![image](https://github.com/ImageMagick/ImageMagick/assets/3017827/560c1722-9e58-464d-ac81-8d5e03982435)

Now I have a few questions. I still have little dots here and there. Usually 2 or 3 pixels, nothing bigger. I have not been able to remove them. Is there something like "remove all groups smaller than X pixels total"?

Also, with the image above, tesseract gives me U 0175S. Not sure why but it duplicates the last caractere. However if I split the lines in different files (manually), then it works perfectly. Knowing that the text will not always been in the exact same place, is there a way I can still cut this horizontaly in 2 lines dynamically?

Here is another example:
![image](https://github.com/ImageMagick/ImageMagick/assets/3017827/5f3cf4ff-8db7-4ec2-9046-7c6f16000c14)
Ouput from Tesseract:
c€. 0149
WOE>+SP h

If I split into 2 images I will get respectively:
c 0149
and
WOE+SP hk

Which is super good for my need.

Thanks.

**Comment by fmw42** [comment]
You can remove small groups of pixels using -connected-components filtering using the area-threshold define.  See https://imagemagick.org/script/connected-components.php

Here is an example starting with the following image that will process the whole thing with just a threshold and connected components to get rid of regions that are 10 pixels in area or less.

![image](https://github.com/ImageMagick/ImageMagick/assets/13458218/39bdf2ff-26a0-45c0-936b-98b81489d4ea)

```
convert image.png \
-threshold 55% -type bilevel -negate \
-define connected-components:mean-color=true \
-define connected-components:area-threshold=10 \
-connected-components 4 result.png
```

![result](https://github.com/ImageMagick/ImageMagick/assets/13458218/aabdabeb-ea3d-4820-aa4c-6bc87d4c693a)

**Reply by jmspaggi** [reply]
Nice! Thanks @fmw42. Way shorter than what I did! I works with many of the images. However, it's having trouble with that one:
![image](https://github.com/ImageMagick/ImageMagick/assets/3017827/c927cba7-4003-4474-a30b-2da4fa6eee57)
With the command above it's not able anymore to detect the text anymore and the 2nd line is mostly gone (I can live with that...) I'm wondering if there is a way for the threshold to be dynamic. Like, I want to keep 20% of the black pixels or something like that.

For the line cutting I'm thinking about applying something like in that thread: https://www.imagemagick.org/discourse-server/viewtopic.php?t=28666

**Reply by fmw42** [reply]
Change the threshold value -- lower it so that it thresholds on the white text.    If you were on IM 7, you could use -auto-threshold otsu and likely get what you want automatically.

**Reply by jmspaggi** [reply]
Very interesting! Let me see if I'm able to build IM7 and use the auto-threshold. I will report here shortly.

**Comment by fmw42** [comment]
You can try the "magick" version that is a complete portable version so that you do not have to build delegates, if that helps you.  See https://imagemagick.org/script/download.php

**Reply by jmspaggi** [reply]
oh, practical! I didn'T know about that! I'm building the version from the source but I tried with magick in the meantime. The syntax might be a bit different.

With:
![image](https://github.com/ImageMagick/ImageMagick/assets/3017827/887c9f35-e5fa-45f3-bafd-89d1d6676cc7)

Using:
./magick output1.jpg -auto-threshold -negate -define connected-components:mean-color=true -define connected-components:area-threshold=10 -connected-components 4 output5.jpg

I get:
![image](https://github.com/ImageMagick/ImageMagick/assets/3017827/d547d35f-07bf-4b2e-8b2f-df8869bfd991)

Because it didn't like it I removed the bilevel option:
magick: unable to open image 'bilevel': Aucun fichier ou dossier de ce type @ error/blob.c/OpenBlob/3571.
